### DISCUSSION TITLE
Can we enrich the type of e.g. dtObstacleRef as `{ ptr: number}` to provide keys?

### DISCUSSION THREAD

**Discussion by rob-myers** [isaac-mason/recast-navigation-js#297] [General]
Hi again,

You provide type `ObstacleRef` which resolves as `class dtObstacleRef {}`.
However, values possess a numeric property `ptr`.

Can we use this number to identity obstacles e.g. when rendering a list of them?

**Comment by isaac-mason** [comment]
Hi, you may be able to, but I'd need to check the source code to see if refs get reused on removing then adding obstacles.

As an alternative - in the js wrapper, or in user land, we can create a unqiue id based on object identity.

```ts
const objectToId = new WeakMap<any, number>()

let nextId = 0

export const id = (object: any) => {
  const id = objectToId.get(object)
  if (id !== undefined) return id

  objectToId.set(object, nextId)
  return nextId++
}
```
